File: prms-web/src/app/pages/management/users/services/user.service.ts
import { Injectable } from '@angular/core';
import { BaseService } from '../../../../core/services/base-service';
import {
  CurrentUserProfilePayload,
  UserCreatePayload,
  UserUpdatePayload,
} from '../models/user.model';
import { UserSearchRequest } from '../models/user-search.request';

@Injectable({
  providedIn: 'root',
})
export class UserService {
  private readonly apiUrl = '/api/v1/users';

  constructor(private base: BaseService) {}

  getCurrent() {
    return this.base.get(`${this.apiUrl}/current`);
  }

  updateCurrentProfile(request: CurrentUserProfilePayload) {
    return this.base.patch(`${this.apiUrl}/current/profile`, request);
  }

  updateCurrentPassword(currentPassword: string, newPassword: string) {
    return this.base.patch(`${this.apiUrl}/current/password`, { currentPassword, newPassword });
  }

  getPage(request: UserSearchRequest) {
    return this.base.post(this.apiUrl + '/page', request);
  }

  getById(id: string) {
    return this.base.get(`${this.apiUrl}/${id}`);
  }

  create(request: UserCreatePayload) {
    return this.base.post(this.apiUrl, request);
  }

  update(id: string, request: UserUpdatePayload) {
    return this.base.put(`${this.apiUrl}/${id}`, request);
  }

  delete(id: string) {
    return this.base.delete(`${this.apiUrl}/${id}`);
  }

  updatePassword(id: string, newPassword: string) {
    return this.base.patch(`${this.apiUrl}/${id}/password`, { newPassword });
  }

  getSkills(id: string) {
    return this.base.get(`${this.apiUrl}/${id}/skills`);
  }

  updateSkills(id: string, items: any[]) {
    return this.base.put(`${this.apiUrl}/${id}/skills`, { items });
  }
}

File: src/main/java/com/tranhuudat/prms/service/UserService.java
package com.tranhuudat.prms.service;

import com.tranhuudat.prms.dto.BaseResponse;
import com.tranhuudat.prms.dto.user.CurrentUserPasswordUpdateRequest;
import com.tranhuudat.prms.dto.user.CurrentUserProfileUpdateRequest;
import com.tranhuudat.prms.dto.user.UserCreateRequest;
import com.tranhuudat.prms.dto.user.UserPasswordUpdateRequest;
import com.tranhuudat.prms.dto.user.UserSearchRequest;
import com.tranhuudat.prms.dto.user.UserUpdateRequest;
import com.tranhuudat.prms.dto.user.UserSkillsUpdateRequest;

import java.util.UUID;

/**
 * @author DatNuclear 04/05/2026 09:00 PM
 * @project prms
 * @package com.tranhuudat.prms.service
 */
public interface UserService {
    BaseResponse getCurrentUser();
    BaseResponse getById(UUID id);
    BaseResponse getPage(UserSearchRequest request);
    BaseResponse create(UserCreateRequest request);
    BaseResponse update(UUID id, UserUpdateRequest request);
    BaseResponse delete(UUID id);
    BaseResponse updatePassword(UUID id, UserPasswordUpdateRequest request);

    BaseResponse updateCurrentUserProfile(CurrentUserProfileUpdateRequest request);

    BaseResponse updateCurrentUserPassword(CurrentUserPasswordUpdateRequest request);

    BaseResponse getSkills(UUID id);

    BaseResponse updateSkills(UUID id, UserSkillsUpdateRequest request);
}
